wen-restart/src/wen_restart.rs

```diff
@@ -1,58 +1,210 @@
 //! The `wen-restart` module handles automatic repair during a cluster restart
 
 use {
-    crate::solana::wen_restart_proto::{
-        MyLastVotedForkSlots, State as RestartState, WenRestartProgress,
+    crate::{
+        last_voted_fork_slots_aggregate::LastVotedForkSlotsAggregate,
+        solana::wen_restart_proto::{
+            LastVotedForkSlotsAggregateRecord, LastVotedForkSlotsRecord, State as RestartState,
+            WenRestartProgress,
+        },
     },
     log::*,
     prost::Message,
-    solana_gossip::{cluster_info::ClusterInfo, epoch_slots::MAX_SLOTS_PER_ENTRY},
+    solana_gossip::{
+        cluster_info::{ClusterInfo, GOSSIP_SLEEP_MILLIS},
+        epoch_slots::MAX_SLOTS_PER_ENTRY,
+    },
     solana_ledger::{ancestor_iterator::AncestorIterator, blockstore::Blockstore},
+    solana_program::{clock::Slot, hash::Hash},
+    solana_runtime::bank_forks::BankForks,
+    solana_sdk::timing::timestamp,
     solana_vote_program::vote_state::VoteTransaction,
     std::{
-        fs::File,
-        io::{Error, Write},
+        collections::{HashMap, HashSet},
+        fs::{read, File},
+        io::{Cursor, Error, Write},
         path::PathBuf,
-        sync::Arc,
+        str::FromStr,
+        sync::{Arc, RwLock},
+        thread::sleep,
+        time::Duration,
     },
 };
 
-pub fn wait_for_wen_restart(
-    wen_restart_path: &PathBuf,
+// If >42% of the validators have this block, repair this block locally.
+const REPAIR_THRESHOLD: f64 = 0.42;
+// Wait until 80% responded before proceeding to next stage.
+const RESTART_SUPERMAJORITY_THRESHOLD: f64 = 0.80;
```

Comment:
nit: can reuse the existing constant `WAIT_FOR_SUPERMAJORITY_THRESHOLD_PERCENT`

Reply:
Changed.